code/crates/test/app/src/state.rs

```diff
@@ -0,0 +1,385 @@
+        store: Store,
+        signing_provider: Ed25519Provider,
+    ) -> Self {
+        Self {
+            ctx,
+            config,
+            address,
+            store,
+            signing_provider,
+            current_height: height,
+            current_round: Round::new(0),
+            current_proposer: None,
+            stream_id: 0,
+            streams_map: PartStreamsMap::new(),
+            rng: StdRng::seed_from_u64(seed_from_address(&address)),
+            peers: HashSet::new(),
+        }
+    }
+
+    /// Returns the earliest height available in the state
+    pub async fn get_earliest_height(&self) -> Height {
+        self.store
+            .min_decided_value_height()
+            .await
+            .unwrap_or_default()
+    }
+
+    /// Processes and adds a new proposal to the state if it's valid
+    /// Returns Some(ProposedValue) if the proposal was accepted, None otherwise
+    pub async fn received_proposal_part(
+        &mut self,
+        from: PeerId,
+        part: StreamMessage<ProposalPart>,
+    ) -> eyre::Result<Option<ProposedValue<TestContext>>> {
+        let sequence = part.sequence;
+
+        // Check if we have a full proposal
+        let Some(parts) = self.streams_map.insert(from, part) else {
+            return Ok(None);
+        };
+
+        // Check if the proposal is outdated
+        if parts.height < self.current_height {
+            debug!(
+                height = %self.current_height,
+                round = %self.current_round,
+                part.height = %parts.height,
+                part.round = %parts.round,
+                part.sequence = %sequence,
+                "Received outdated proposal part, ignoring"
+            );
+
+            return Ok(None);
+        }
+
+        // Re-assemble the proposal from its parts
+        let value = assemble_value_from_parts(parts);
+
+        self.store.store_undecided_proposal(value.clone()).await?;
+
+        Ok(Some(value))
+    }
+
+    /// Retrieves a decided block at the given height
+    pub async fn get_decided_value(&self, height: Height) -> Option<DecidedValue> {
+        self.store.get_decided_value(height).await.ok().flatten()
+    }
+
+    /// Commits a value with the given certificate, updating internal state
+    /// and moving to the next height
+    pub async fn commit(
+        &mut self,
+        certificate: CommitCertificate<TestContext>,
+    ) -> eyre::Result<()> {
+        let Ok(Some(proposal)) = self
+            .store
+            .get_undecided_proposal(certificate.height, certificate.round)
+            .await
+        else {
+            return Err(eyre!(
+                "Trying to commit a value at height {} and round {} that is not decided: {}",
+                certificate.height,
+                certificate.round,
+                certificate.value_id
+            ));
+        };
+
+        self.store
+            .store_decided_value(&certificate, proposal.value)
+            .await?;
+
```

Comment:
Maybe we should remove value from the undecided table.

Reply:
https://github.com/informalsystems/malachite/pull/747/commits/0971317dbc1c0f0276bbf891bc45c4f854c1c4a8